pkg/applications/providers/source/git.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2022 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"context"
+	"fmt"
+	"path"
+
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	gossh "golang.org/x/crypto/ssh"
+
+	appskubermaticv1 "k8c.io/kubermatic/v2/pkg/apis/apps.kubermatic/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GitSource download the application's source from a git repository.
+type GitSource struct {
+	Ctx context.Context
+
+	// Client to seed cluster.
+	Client ctrlruntimeclient.Client
+
+	Source *appskubermaticv1.GitSource
+
+	// Namespace where credential secrets are stored.
+	SecretNamespace string
+}
+
+// DownloadSource clone the repository into destination and return the full path to the application's sources.
+func (g GitSource) DownloadSource(destination string) (string, error) {
+	auth, err := g.authFromCredentials()
+	if err != nil {
+		return "", err
+	}
+
+	checkoutStrategy := g.getCheckoutStrategy()
+	if err := checkoutStrategy(g.Ctx, destination, g.Source, auth); err != nil {
```

Comment:
nit

`checkoutStrategy(g.Ctx, destination, g.Source, auth)` looks a bit misleading. Since we are actually performing a `git checkout` here. Maybe we can do something like `checkout := g.getCheckoutStrategy()`

P.s. My concern here is just code coherence, nothing logical or syntactical.